java-manta-client-unshaded/src/main/java/com/joyent/manta/config/AuthAwareConfigContext.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2017, Joyent, Inc. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.joyent.manta.config;
+
+import com.joyent.http.signature.Signer;
+import com.joyent.http.signature.ThreadLocalSigner;
+import com.joyent.http.signature.apache.httpclient.HttpSignatureAuthScheme;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.http.auth.Credentials;
+import org.apache.http.auth.UsernamePasswordCredentials;
+
+import java.security.KeyPair;
+import java.util.Objects;
+
+/**
+ * Object for tracking configuration changes related to authentication and recreating dependent objects as needed.
+ * It combines the users' configuration with the derived runtime objects needed to authenticate requests.
+ * objects like the {@link ThreadLocalSigner} which needs careful lifecycle management.
+ *
+ * @author <a href="https://github.com/tjcelaya">Tomas Celayac</a>
+ * @since 3.1.7
+ */
+public class AuthAwareConfigContext
+        extends StandardConfigContext
+        implements AutoCloseable {
+
+    /**
+     * Hashcode of last observed authentication-related configuration values.
+     */
+    private volatile int parametersFingerprint;
+
+    /**
+     * Atomic reference to objects we provide.
+     */
+    private volatile AuthContext authContext;
+
+    /**
+     * Build an AuthAwareConfigContext from an existing {@link ConfigContext}.
+     *
+     * @param config configuration context from which to extract values
+     */
+    public AuthAwareConfigContext(final ConfigContext config) {
+        overwriteWithContext(config);
+        reload();
+    }
+
+    /**
+     * Check the configuration for authentication-related changes. Clean up old authentication objects which might
+     * still exist and load new instances.
+     *
+     * @return this after reload
+     */
+    public synchronized AuthAwareConfigContext reload() {
```

Comment:
I've been staring a this `synchronized` with doubt as to its necessity and I'm afraid I have another concern.  If thread A calls `setEncryptionPrivateKeyPath`, thread B calls `setMantaKeyId`, and thread C calls `reload()` I *think* we have a potential problem since this `synchronized` is effectively protecting the read of all fields, but not the writes.  I don't have my copy of JCIP on hand to verify with the right term.  Please let me know if that is confusing.

One possibility to avoid a lot of what would usually be unnecessary synchronization would be to create a ConfigContext that uses a `ReentrantReadWriteLock` and then inherent from that.

Reply:
You're right, the modification of the bag of parameter values in the config context is not totally threadsafe. I didn't override all the setters in an effort to keep the PR easy to review, and while I acknowledge that concurrent writes to the `ConfigContext` are a possibility I wanted to avoid complicating the general usage to accommodate an edge case within an edge case. I'll experiment with a `ReentrantReadWriteLock` and see how much that complicates `AuthAwareConfigContext`.

Reply:
After further consideration don't really want to go as far as wrapping all the config values in a lock since some of them are accessed extremely often. Would documenting the thread-unsafety of concurrent modifications to `AuthAwareConfigContext` be adequate? Even with locking there's no way to guarantee that a single consistent version of a config will be used for the entire request since the client may read from the `ConfigContext` anytime while executing a request (e.g. verifyUploads is checked late in the request cycle relative to username)

Reply:
One thought that might be out there - you could have all of the setters and getters wrap to operations that store their data in a `ConcurrentHashMap` instance. By doing so, you should get an optimistic locking implementation by default. In fact, you may be able to inherit from `MapConfigContext` and just supply it with the `ConcurrentHashMap`.

Reply:
The issue @cburroughs is getting at is more about synchronizing the creation of derived objects with updates to the underlying configuration itself. Even if the parameters themselves were allowed to be accessed concurrently we'd still need a lock that around parameters being updated while we're inside the `reload` method.